server/src/main/java/org/jboss/as/server/suspend/SuspendController.java

```diff
@@ -143,56 +143,57 @@ private Iterator<List<SuspendableActivity>> resumeIterator() {
         CompletableFuture<Void> result = new CompletableFuture<>();
-        // Counter used to determine when to complete final stage
-        AtomicInteger counter = new AtomicInteger(this.activityGroups.size());
+        // Iterate over activity groups (in the order dictated by the caller)
+        Iterator<List<SuspendableActivity>> groups = activityGroups.iterator();
         BiConsumer<Void, Throwable> groupCompleter = new BiConsumer<>() {
             @Override
             public void accept(Void ignore, Throwable exception) {
                 if (exception != null) {
                     result.completeExceptionally(exception);
-                } else if (counter.decrementAndGet() == 0) {
+                } else if (!groups.hasNext()) {
+                    // No more groups
                     result.complete(null);
-                }
-            }
-        };
-        // Iterate over activity groups (in the order dictated by the caller)
-        for (List<SuspendableActivity> group : activityGroups) {
-            List<SuspendableActivity> activities = List.copyOf(group);
-            if (activities.isEmpty()) {
-                // There are no activities for this group, complete immediately
-                groupCompleter.accept(null, null);
-            } else {
-                // Stage for this group will complete after all activity stages complete
-                CompletableFuture<Void> groupStage = new CompletableFuture<>();
-                groupStage.whenComplete(groupCompleter);
-                // Counter used to determine when to complete group stage
-                AtomicInteger groupCounter = new AtomicInteger(activities.size());
-                for (SuspendableActivity activity : activities) {
-                    BiConsumer<Void, Throwable> activityCompleter = new BiConsumer<>() {
-                        @Override
-                        public void accept(Void ignore, Throwable exception) {
-                            if (exception != null) {
-                                try {
-                                    exceptionHandler.accept(activity, exception);
-                                } finally {
-                                    groupStage.completeExceptionally(exception);
+                } else {
+                    // Create stage for next group
+                    List<SuspendableActivity> activities = List.copyOf(groups.next());
+                    CompletableFuture<Void> groupStage = new CompletableFuture<>();
+                    groupStage.whenComplete(this);
+                    if (activities.isEmpty()) {
+                        // No activities, complete immediately
+                        groupStage.complete(null);
+                    } else {
+                        // Counter used to determine when to complete group stage
+                        AtomicInteger groupCounter = new AtomicInteger(activities.size());
```

Comment:
Nitpick: This should be named `activityCounter` instead `groupCounter` due to what it represents.

Reply:
Let me fix that - surely that's a potential source of confusion.